---
title: 'BIOL119-2024: Gene Expression Analysis'
author: 'Author: Your Name'
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: yes
    toc_float:
      collapsed: yes
      smooth_scroll: yes
    toc_depth: 3
    fig_caption: yes
    code_folding: show
    number_sections: yes
  pdf_document:
    toc: yes
    toc_depth: '3'
  word_document:
    toc: yes
    toc_depth: '3'
fontsize: 14pt
bibliography: bibtex.bib
type: docs
weight: 304
editor_options: 
  chunk_output_type: console
---

<!---
- Compile from command-line
Rscript -e "rmarkdown::render('exprAnalyisDemo.Rmd', c('html_document'), clean=FALSE)"
-->

# Get started 

The following provides brief instructions how to access RStudio Cloud (or
R/RStudio), clone the pre-generated project for this exercise, and run
the provided analysis code. The latter automatically downloads genome-wide expression data
from NCBI's GEO database, and then identifies:

+ differentially expressed genes (DEGs)
+ enriched GO terms
+ clusters of co-expressed genes

## Initialize project 

To get started with [RStudio Cloud](https://rstudio.cloud/), users can follow the instructions below.
Instead of RStudio Cloud, more advanced users can install everything on their local computer
by first installing R from [CRAN](https://cran.r-project.org), and then the free version
of RStudio from [here](https://www.rstudio.com/products/rstudio/download/). Since the
usage of RStudio Cloud does not require any local software installs, this
method is recommended for users who are new to R. For more advanced users the
local install option is less restrictive and more flexible. In both cases the
user interfaces are extremely similar and the way to run R code is nearly
identical. 

### Simple link-based approach
Note, students who are entirely new to R should use this approach (1.1.1) instead of (1.1.2):

1. Create a free personal account on Posit (RStudio) Cloud
   [here](https://posit.cloud/). 
2. Next click the <span style="background-color: #f1c40f;">Posit (RStudio) Cloud assignment link</span></a> provided under the same name
   coding on the corresponding assignment page on Canvas
   [here](https://elearn.ucr.edu/courses/147748/assignments/750221). Note
   the latter page is private to this class.
3. Start the pregenerated assignment by clicking the `BIOL119_Expression_Analysis`. This will create a copy of the assignment in a student's account.  
4. Continue under section 1.2 and skip R package install under section 2.

### Clone assignment from GitHub
This alternative approach is for slightly more advanced R users.

1. Create a free personal account on RStudio Cloud
   [here](https://rstudio.cloud/). 
2. Go to RStudio Cloud [here](https://rstudio.cloud/projects/)
3. Select under blue _New Project_ menu: `New Project from Git Repository`
4. Provide this URL to clone BIOL119 repos: [https://github.com/tgirke/BIOL119.git](https://github.com/tgirke/BIOL119.git) 
5. Continue under section 1.2. Note, the initial R package installs under section 2 will be 
   required for this approach.

## Load R code and run it

The R code along with the human readable text in this document can be opened and executed as follows:

1. Open source `Rmd` file for this document by selecting in top menu: `File` ->
   `Open File` -> `exprAnalyisDemo.Rmd`. Alternatively, one can click the name of this 
   file in Posit's (RStudio's) file browser, usually located in the window on the bottom right. Either option
   will open the `Rmd` in the code editor window usually located in the top left window.
2. Execute code of the `Rmd` file line by line by placing the cursor on a line of R code 
   and then pressing on the keyboard `Ctrl + Enter`. This will send the corresponding code line
   to the R console (located on bottom left) and execute (run) the code. It is important to execute the
   lines in the order given in the `Rmd` file. Note, `Rmd` files are special `R` files that contain 
   both human readable text in markdown format as well as R code located in so called code chunk boxes 
   that are initialized and terminated with a tag line starting with three backtick (\`) characters. If an 
   `R` script instead of an `Rmd` script is used then everything works the same. However, 
   in an `R` script human readable text has to be initialized on each line with a comment (`#`) sign.

Alternatively, one can execute all code in this document at once simply by
pressing the triangle next to the Knit button above the code editor window in
RStudio, and then selecting `Knit to HTML`. This will not only execute the
code, but also regenerate this HTML document and update all output, tables and
figures generated by the evaluated R code chunks accordingly.  Similarly, one
can generate a PDF document instead. The environment that makes this possible
is called [R Markdown](https://rmarkdown.rstudio.com/).

# Install required packages

The following will install all packages required for this exercise. A fresh
install into a new R project will take about 2-3 minutes. This install only
needs to be done once, meaning users want to skip this step after restarting R
or rerunning the analysis. Users who have loaded the pre-configured assignment
project into their RStudio Cloud account under section 1.1.1 can skip this
initial package install since a classroom project comes with all packages 
preinstalled.

```{r package_installs, eval=FALSE, message=FALSE, warning=FALSE}
if(!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
BiocManager::install(c("GEOquery", "limma", "clusterProfiler", "org.Hs.eg.db", "enrichplot", "pheatmap", "RColorBrewer", "rmarkdown"))
```

# Import data from GEO

The following uses the `GEOquery` package to automatically download from GEO a
expression data set that involves a chemical treatment. The specific data set
chosen for this exercise is experiment
[GDS2778](https://www.ncbi.nlm.nih.gov/sites/GDSbrowser?acc=GDS2778), which is the 
same one used for the online exercises of computer section SA5 of BIOL119. This
Affymetrix gene expression profiling experiment from Gillis _et al._
[-@Gillis2007-qk] aims to study the effect of the air pollutant
1,2,4-benzenetriol on blood cells to provide insight into the molecular basis
of benzene cytotoxicity.  More detailed annotation information about this
experiment is available under Reference Series
[GSE7664](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE7664). 

```{r import_data_from_geo, eval=TRUE, message=FALSE}
library(GEOquery) 
gds <- getGEO("GDS2778") # Download GDS2778 data from GEO
eset <- GDS2eSet(gds)
```

The experimental design information can be returned with the `pData` function. This includes
sample names, and control, treatment and replicate information.

```{r inspect_data_from_geo_1, eval=TRUE, message=FALSE}
pData(eset)[ ,1:2] 
```
The expression data can be returned with the `exprs` function (here for first 4
rows). In this case the authors have normalized the data already with the RMA
method [@Gautier2004-tk]. This is why the normalization step is skipped here.
Note, RMA returns log2 transformed values which is the expected input format for the
downstream DEG analysis with `limma`. Users who are interested in
learning to perform the normalization themselves want to look into the `rma`
function that is part of the `affy` package from Bioconductor (see
[here](https://bioconductor.org/packages/release/bioc/html/affy.html)).


```{r inspect_data_from_geo_2, eval=TRUE, message=FALSE}
dim(exprs(eset))
exprs(eset)[1:4, ] 
```

# DEG analysis with Limma

The following identifies differentially expressed genes with the `limma`
package [@Ritchie2015-rl]. This includes the following steps: (1) creation of a
design matrix for this data set with `model.matrix`; (2) fitting of a linear
model for each gene based on the given series of arrays with `lmFit`; (3)
creation of a contrast matrix defining the sample comparisons with
`makeContrasts`; (4) estimation of coefficients and standard errors for the
chosen contrast(s) with `contrast.fit`; and (5) computation of moderated
t-statistics and log-odds of differential expression by empirical Bayes
shrinkage of the standard errors towards a common value using `eBayes`.
Subsequently, the final results can be accessed with the `topTable` function.

```{r deg_analysis, eval=TRUE, message=FALSE}
library(limma) # Loads limma library.
targets <- pData(eset)
design <- model.matrix(~ -1+factor(as.character(targets$agent)))
colnames(design) <- c("Treatment", "Control") 
fit <- lmFit(eset, design) 
contrast.matrix <- makeContrasts(Control-Treatment, levels=design)
fit2 <- contrasts.fit(fit, contrast.matrix) 
fit2 <- eBayes(fit2)
deg_df <- topTable(fit2, coef=1, adjust="fdr", sort.by="B", number=Inf)
deg_df <- deg_df[!is.na(deg_df$adj.P.Val),] # Removes rows with NA values
head(deg_df)[,22:27]
```

## DEG filtering

The `deg_df` table (here `data.frame` object) generated above contains the DEG
results for all genes without applying any filtering. To identify how many genes
meet certain filter criteria (cutoffs), one can first check the total number of
rows in this table where each row corresponds to a probe set of a gene. The
`nrow` function can be used to return the number of rows in this table (here
data.frame). Note, most probe sets (IDs in first column or row name slot) have 
a one-to-one relationship to the corresponding target genes. However, in some 
cases there are duplications. For simplicity this exercise assumes a one-to-one 
relationship among the two. 

```{r deg_filter_1, eval=TRUE, message=FALSE}
nr <- nrow(deg_df)
nr
```

As shown above the total number of rows or genes (probe sets) in `deg_df` is `r prettyNum(nr, big.mark=",", scientific=FALSE)`. 

To obtain the number of genes that are differentially expressed based on an adjusted 
p-value (`adj.P.Val`) cutoff of $\le 0.01$, one can use the following filtering syntax.

```{r deg_filter_2, eval=TRUE, message=FALSE}
pf <- nrow(deg_df[deg_df$adj.P.Val <= 0.01, ])
pf
```

The number of DEGs passing this filter is `r prettyNum(pf, big.mark=",", scientific=FALSE)`.

To obtain the number of genes with a fold change (`logFC`) of at least 2, 
one can use the following filtering syntax. Note, a 2-fold change or higher corresponds
on a log2 scale (here `logFC`) to: $\ge 1$ OR $\le -1$.

```{r deg_filter_3, eval=TRUE, message=FALSE}
fcf <- nrow(deg_df[deg_df$logFC >= 1 | deg_df$logFC <= -1, ])
fcf
```

The number of DEGs passing this filter is `r prettyNum(fcf, big.mark=",", scientific=FALSE)`.

To apply a combinatorial filter for both fold changes of at least 2 and an adjusted p-value
of $\le 0.01$, one can use the following syntax. For readability the individual filtering results 
(here `pf_log` and `fcf_log`) are generated on separate lines and then combined (under `comb_filter`).

```{r deg_filter_4, eval=TRUE, message=FALSE}
pf_log <- deg_df$adj.P.Val <= 0.01
fcf_log <- deg_df$logFC >= 1 | deg_df$logFC <= -1
comb_filter <- pf_log & fcf_log
combf <- nrow(deg_df[comb_filter, ])
combf
```

The number of DEGs passing this filter is `r prettyNum(combf, big.mark=",", scientific=FALSE)`.

# Enrichment analysis of GO terms 

The following performs over-representation analysis (ORA) of GO terms using functions from 
the `clusterProfiler` package [@Yu2012-hr].

## Prepare input

For the GO term enrichment analysis the DEGs are filtered using a adjusted
p-value of $\le 0.05$. Subsequently, the gene identifiers (Entrez IDs) are
stored in a character vector named `ids`. 

```{r overrepresentation_analysis, eval=TRUE, message=FALSE}
cutoff <- 0.05 # Cutoff to use for filtering on adjusted p-value (FDR)
ids <- deg_df[deg_df$adj.P.Val <= cutoff, "Gene.ID"]
ids <- ids[!grepl("/", ids)] # Removes ambiguous mappings
ids <- ids[nchar(ids)!=0] # Removes empty slots
ids # Prints gene IDs
```

## Enrichment of MF terms

The following uses the `clusterProfiler` package to perform over-represention
analysis (ORA) using the Molecular Function (MF) Gene Ontology as annotation system. 
To return only significant results a p-value cutoff of $\le 0.1$ is applied.

```{r overrepresentation_analysis_mf, eval=TRUE, message=FALSE}
library(clusterProfiler); library(org.Hs.eg.db); library(enrichplot)
ego_mf <- enrichGO(gene=ids, OrgDb=org.Hs.eg.db, ont="MF", pAdjustMethod="BH", pvalueCutoff=0.1, readable=TRUE)
dim(ego_mf) # Returns number of rows and columns in result table
head(ego_mf) # Returns first six rows to inspect results
```

## Visualization of MF result

The `barplot` plots the top scoring GO terms (here 10) in form of a bar plot. To plot 
the tree structure of the corresponding DAG, the `goplot` function can be used. 

```{r plot_enrichment_mf, eval=TRUE, message=FALSE}
barplot(ego_mf, showCategory=10)
# goplot(ego_mf)
```

## Enrichment of BP terms

Same as above but for Biological Process (BP) Gene Ontology.

```{r overrepresentation_analysis_bp, eval=TRUE, message=FALSE}
ego_bp <- enrichGO(gene=ids, OrgDb=org.Hs.eg.db, ont="BP", pAdjustMethod="BH", pvalueCutoff=0.1, readable=TRUE)
dim(ego_bp)
head(ego_bp)
```

## Visualization of BP result

Same bar plot as above but for Biological Process (BP) Gene Ontology.

```{r plot_enrichment_bp, eval=TRUE, message=FALSE}
barplot(ego_bp, showCategory=10)
```

# Clustering

The following uses DEGs passing an adjusted p-value cutoff of $\le 0.05$ to
subset the gene expression matrix imported from GEO. The subsetted matrix is then used
for hierarchical clustering. The result is visualized in form of a heatmap
where the rows and columns are sorted by the gene- and sample-wise hierarchical
clustering dendrograms, respectively. The actual expression values are represented 
in the heatmap using a custom color scheme. In this case, row-wise scaling (see `scale=row`) 
is applied to maximize the number of colors used for visualizing the expression profile for 
each gene. Users interested in learning more about clustering in R can consult this [tutorial](https://girke.bioinformatics.ucr.edu/GEN242/tutorials/rclustering/rclustering/).


```{r clustering, eval=TRUE, message=FALSE}
library(pheatmap); library("RColorBrewer")
cutoff <- 0.05 # Cutoff to use for filtering on adjusted p-value (FDR)
affy_ids <- row.names(deg_df[deg_df$adj.P.Val <= cutoff, ])
deg_ma <- exprs(eset)[affy_ids, ] 
pheatmap(deg_ma, scale="row", color=brewer.pal(9, "Blues"))
```

# Session Info

```{r sessionInfo}
sessionInfo()
```

# References
