---
title: 'Phylogenetic Analysis in R'
subtitle: 'Part 2 of Discussion Section SA4 of BIOL119-2025'
author: 'Author: Your Name'
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: yes
    toc_float:
      collapsed: yes
      smooth_scroll: yes
    toc_depth: 3
    fig_caption: yes
    code_folding: show
    number_sections: yes
  pdf_document:
    toc: yes
    toc_depth: '3'
  word_document:
    toc: yes
    toc_depth: '3'
fontsize: 14pt
bibliography: bibtex.bib
type: docs
weight: 304
editor_options: 
  chunk_output_type: console
---

<!---
- Compile from command-line
Rscript -e "rmarkdown::render('phyloAnalysis.Rmd', c('html_document'), clean=FALSE)"
-->

# Introduction

This tutorial covers a basic phylogenetic analysis in R including the following steps. It starts with an external website query, followed by all subsequent steps in R.

1. NCBI query to retrieve sequence IDs 
2. Sequence batch download
3. Cration of multiple sequence alignment
4. Building phylogentic tree with bootstrap analysis
5. Tree visualization


# Creation of Input File

The following queries NCBI's protein database to retrieve the protein IDs of all Cytochrome P450 Hydroxilases from human that are present in UniProt.

 - Go to **[NCBI Protein](http://www.ncbi.nlm.nih.gov/protein)**, and run the following query: `P450 & hydroxylase & human [organism]`
 - Select under “Source databases” **UniProtKB/SwissProt** as database.
 - Save a *list of identifiers* (**GI List**) to your desktop via the "Send to" File drop down menu.

Note, the abbreviation "GI" often stands for "Gene ID".  However, these are *not* Entrez Gene IDs.  Instead, it is more appropriate to consider these "Protein IDs."

Because this is a deviation from SA2 (creation of identifier list *instead* of FASTA file), this screenshot helps illustrate the novel step:

![Identifier Download](data/phyloAnalysis/NCBI_Protein.png "Downloading List of Identifiers")

The result is a `sequence.gi` file.

# Import Libraries
R packages required for this exercise will be loaded with the `library` command. To suppress verbose messages during the package loading process, the `suppressPackageStartupMessages` is used.

```{r example1, eval=TRUE, message=FALSE}
suppressPackageStartupMessages({
	library(rentrez)
	library(Biostrings)
	library(org.Hs.eg.db)
	library(seqinr)#to create intermediate FASTA file
	library(msa)
	library(ape)
	library(pheatmap)
})
```

Custom functions can be loaded with the ``source` command.
```{r example1_fct, eval=TRUE, message=FALSE}
source("phyloAnalysis_Fct.R")
```


# Get IDs 

## Gene IDs

The `rentrez` package in R is utilized to import the sequences corresponding to the IDs obtained in the previous step.

```{r example2, eval=TRUE, message=FALSE}
result_ID_file <- "data/phyloAnalysis/sequence.gi"
linked_gene_IDs <- geneID_wrapper(result_ID_file)
```

## Gene Symbols 

The corresponding gene symbols can be obtained by querying the `org.Hs.eg.db` data package.

```{r example3, eval=TRUE, message=FALSE}
gene_symbols <- mapIds( x = org.Hs.eg.db, keys = linked_gene_IDs,
					column = "SYMBOL", keytype = "ENTREZID")
					
FASTA_file <- "rentrez-P450_UniProtKB_SwissProt-gene_symbol.fasta" # for OUTPUT, not input
FASTA_symbol_wrapper(result_ID_file, gene_symbols, FASTA_file)
```

# Multiple Sequence Alignment

The `msa` package in Bioconductor [@bodenhofer2015msa] offers access to three multiple alignment software tools: ClustalW [@thompson1994clustal], Clustal Omega [@sievers2011fast], and MUSCLE [@edgar2004muscle].

Multiple Sequence Alignments (MSA) are commonly represented using the "clustal" alignment format or the gapped FASTA format. This document utilizes Clustal Omega to generate the MSA for the downloaded P450 sequences due to its more recent implementation.

For an alternative approach, the web version of Clustal Omega can be accessed at https://www.ebi.ac.uk/jdispatcher/msa/clustalo.

However, using the provided R code offers several benefits beyond reproducibility. These include the addition of bootstrap confidence and increased flexibility, such as the ability to generate heatmaps from distance values, as demonstrated in a later section.


```{r example4, eval=TRUE, message=FALSE}
P450_SeqObj <- readAAStringSet(FASTA_file)

msa_ClustalOmega <- msa(P450_SeqObj, method = "ClustalOmega")

# View Multiple Alignment Near Invariant Cysteine
msa_basename <- "rentrez-msa_ClustalOmega-P452_UniProtKB_SwissProt.msaPrettyPrint.pdf" #please note the folder that is saved for GitHub visualization
msa_pdf <- paste("./data/phyloAnalysis",msa_basename,sep="/") #please note the folder that is saved for GitHub visualization
msaPrettyPrint(msa_ClustalOmega, file = msa_basename,
			y=c(540, 640), output="pdf", showNames="left", 
			logoColors="rasmol", showLogo="top",
			shadingMode="identical", shadingModeArg =c(50,90), shadingColors = "grays",
			askForOverwrite=FALSE)
file_copy(msa_basename, msa_pdf) #make sure that the PDF is saved in the intended location

msa_ClustalOmega.ape <- msaConvert(msa_ClustalOmega, type="ape::AAbin") #conversion of R object for next step
```

![P450 Visualization Created Using `msaPrettyPrint`](./data/phyloAnalysis/rentrez-msa_ClustalOmega-P452_UniProtKB_SwissProt.msaPrettyPrint.pdf){width=100%}

# Visualize Distance Matrix

A heatmap can be used to visualize a distance matrix. The `dist.aa` function from
the `pheatmap` package provides an example of this. In the resulting plot, colors
represent the distance values, and the order of the rows and columns is
adjusted according to distance trees.

```{r example5, eval=TRUE, message=FALSE}
msa_ClustalOmega.dist <- dist.aa(msa_ClustalOmega.ape, pairwise.deletion = FALSE, scaled = FALSE)

pheatmap(as.matrix(msa_ClustalOmega.dist))
```

# Phylogenetic Tree with Bootstrap

Phylogenetic analysis, as performed by the ape Bioconductor package
[@paradis2004ape], offers various methods for tree construction. This package
can build phylogenetic trees from a distance matrix using either
Neighbor-Joining (NJ) [@trees1987neighbor] or Unweighted Pair Group Method with
Arithmetic Averages (UPGMA) [@sokal1958statistical].

While this example focuses on amino acid divergence, nucleotide divergence can
also be analyzed. For nucleotide sequences, Maximum Likelihood (ML) can be
employed to compare substitution models and generate the most probable tree
[@felsenstein1981evolutionary]. Although ML trees can also be constructed for
amino acid sequences, the underlying model would not be based on nucleotide
substitution rates. Maximum Parsimony is another available method for
phylogenetic analysis.


```{r example6, eval=TRUE, message=FALSE}
ape_tree_file <- "rentrez-msa_ClustalOmega-P452_UniProtKB_SwissProt.ape_tree.bootstrap-gene_symbol.png" #for OUTPUT, not input

tree_function <- function(x) nj(dist.aa(x))
bootstrap_n <- 100 #should really use at least 1000, but we are reducing time and resources

ape_bootstrap_wrapper.clade_only(msa_ClustalOmega.ape, tree_function, bootstrap_n)
```

# Study Material for Homework

## Lecture LA6
 - UPGMA / NJ Implementation
 - UPGMA / NJ Assumptions
 - Bootstrapping Support for Tree Generation

## Procedures from This Assignment (SA4)
 - Basic R Functions / Syntax
 - Descriptive Statistic Calculations in R
 - Knowledge of Commonly Used Functions from R Libraries

# Session Info

```{r sessionInfo}
sessionInfo()
```


# References
